### DISCUSSION TITLE
Performance improvement for functional calling

### DISCUSSION THREAD

**Discussion by HuskyDanny** [microsoft/semantic-kernel#9907] [General]
Hi Team, 

I have a trace monitoring showing 17s spent on a semantic function
![image](https://github.com/user-attachments/assets/4d086f32-8365-4bf7-96dc-b9692cc56aba)
I enabled functional calling, so the back and forth network calls to get the function execution might take the most of the time.

My question is "Is there a way to inject log for the substep of functional calling or a specific feature in smeantic kernel if there is no native one"? 

The reason why I want to do it is I want to find the root cause of 17s and find potential improvement on it based on spans and logs

**Comment by moonbox3** [comment]
@TaoChenOSU do you have thoughts on this?

**Comment by TaoChenOSU** [comment]
Hi @HuskyDanny,

Thank you for your question!

By looking at the timeline you posted, it only shows one function call within `rag_qa_with_context`. I will need to get more context of your scenario to help you find out what is happening within your function calling logic.

To answer your question, by default SK functions already generate a set of logs at different log levels. It's also possible to write more logs. If you are using native functions, you can write logs within them. If you are using prompt functions, you can use a filter to create more logs: https://review.learn.microsoft.com/en-us/semantic-kernel/concepts/enterprise-readiness/filters?branch=main&pivots=programming-language-csharp

Could you please let us know which language you are using?

**Reply by HuskyDanny** [reply]
Thanks for the prompt response, I am using python semantic-kernel==1.7.0. Specifically, with functional calling, there are multiple hops to get the function execution metadata, but the built in logs didn't send them automatically, how can we add spans to these substeps?

**Reply by TaoChenOSU** [reply]
We have a sample here that shows how to set up logs and traces: https://github.com/microsoft/semantic-kernel/tree/main/python/samples/demos/telemetry. Please take a look and feel free to pose further questions.
